File: LeeCode/binaryTreePreorderTraversal.swift
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */
class Solution {
    func preorderTraversal(_ root: TreeNode?) -> [Int] {
        var arr = [Int]()
        preTraversal(root, array: &arr)
        return arr
    }
    func preTraversal(_ root: TreeNode?, array: inout [Int]) {
        guard root != nil else {
            return
        }
        array.append(root!.val)
        preTraversal(root?.left, array: &array)
        preTraversal(root?.right, array: &array)
    }
}

File: LeeCode/binaryTreePreorderTraversal.py

# import stac
# Definition for a binary tree node.


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    # 迭代法先序遍历二叉树
    def preorderTraversal(self, root: TreeNode) -> [int]:
        ret = []
        cache = []
        curr = root
        while curr or len(cache) > 0:
            if curr:
                ret.append(curr.val)
                cache.append(curr)
                curr = curr.left
            else:
                curr = cache.pop()
                curr = curr.right

        return ret

    def preorderTraversal2(self, root: TreeNode) -> [int]:
        result = []
        if root:
            tempArr = []
            tempArr.append(root)
            while len(tempArr) > 0:
                top = tempArr.pop()
                result.append(top.val)
                if top.right:
                    tempArr.append(top.right)
                if top.left:
                    tempArr.append(top.left)
        return result

root = TreeNode(1)
root.left = TreeNode(2)
root.right = TreeNode(3)
root.left.left = TreeNode(4)
root.left.right = TreeNode(5)

print(Solution().preorderTraversal(root))
print(Solution().preorderTraversal2(root))


# test = [1,2,3,4,5]
# print(test.pop())
# print(test)
